File: autoconduct/usage.py
"""Query Claude subscription usage via the OAuth endpoint.

Uses the OAuth access token Claude Code already stores in the macOS
Keychain (service: "Claude Code-credentials"). Read-only; no secrets are
ever written to disk or logs.
"""

from __future__ import annotations

import json
import subprocess
import urllib.request
from dataclasses import dataclass
from datetime import datetime

USAGE_URL = "https://api.anthropic.com/api/oauth/usage"
KEYCHAIN_SERVICE = "Claude Code-credentials"


@dataclass(frozen=True)
class Window:
    """One rate-limit window: percent used and when it resets."""

    utilization: float  # 0..100
    resets_at: datetime | None


@dataclass(frozen=True)
class UsageSnapshot:
    five_hour: Window
    seven_day: Window
    fetched_at: datetime


class UsageError(RuntimeError):
    """Raised when usage cannot be determined. Callers must fail closed."""


def _read_oauth_token() -> str:
    try:
        out = subprocess.run(
            ["security", "find-generic-password", "-s", KEYCHAIN_SERVICE, "-w"],
            capture_output=True,
            text=True,
            check=True,
            timeout=10,
        ).stdout.strip()
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as exc:
        raise UsageError(f"Cannot read Claude credentials from Keychain: {exc}") from exc
    try:
        return json.loads(out)["claudeAiOauth"]["accessToken"]
    except (json.JSONDecodeError, KeyError) as exc:
        raise UsageError("Unexpected Keychain credential format") from exc


def _parse_window(raw: dict | None) -> Window:
    if not raw:
        return Window(utilization=0.0, resets_at=None)
    resets = raw.get("resets_at")
    return Window(
        utilization=float(raw.get("utilization") or 0.0),
        resets_at=datetime.fromisoformat(resets) if resets else None,
    )


def parse_usage(payload: dict, fetched_at: datetime) -> UsageSnapshot:
    """Parse the /api/oauth/usage payload into an immutable snapshot."""
    return UsageSnapshot(
        five_hour=_parse_window(payload.get("five_hour")),
        seven_day=_parse_window(payload.get("seven_day")),
        fetched_at=fetched_at,
    )


def fetch_usage(now: datetime) -> UsageSnapshot:
    token = _read_oauth_token()
    req = urllib.request.Request(
        USAGE_URL,
        headers={
            "Authorization": f"Bearer {token}",
            "anthropic-beta": "oauth-2025-04-20",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            payload = json.loads(resp.read().decode("utf-8"))
    except Exception as exc:  # noqa: BLE001 — network errors must fail closed
        raise UsageError(f"Usage endpoint unreachable: {exc}") from exc
    return parse_usage(payload, fetched_at=now)

File: raycast/src/lib/usage.ts
import { execFile } from "child_process";
import { promisify } from "util";

const execFileAsync = promisify(execFile);

export interface Usage {
  fiveHour: number; // 0..100
  weekly: number;
  fiveHourResetsAt?: string;
  weeklyResetsAt?: string;
}

async function readToken(): Promise<string> {
  const { stdout } = await execFileAsync("/usr/bin/security", [
    "find-generic-password",
    "-s",
    "Claude Code-credentials",
    "-w",
  ]);
  const creds = JSON.parse(stdout) as { claudeAiOauth?: { accessToken?: string } };
  const token = creds.claudeAiOauth?.accessToken;
  if (!token) throw new Error("No Claude credentials found (is Claude Code logged in?)");
  return token;
}

export async function fetchUsage(): Promise<Usage> {
  const token = await readToken();
  const res = await fetch("https://api.anthropic.com/api/oauth/usage", {
    headers: { Authorization: `Bearer ${token}`, "anthropic-beta": "oauth-2025-04-20" },
  });
  if (!res.ok) throw new Error(`Usage endpoint returned ${res.status}`);
  const j = (await res.json()) as {
    five_hour?: { utilization?: number; resets_at?: string };
    seven_day?: { utilization?: number; resets_at?: string };
  };
  return {
    fiveHour: j.five_hour?.utilization ?? 0,
    weekly: j.seven_day?.utilization ?? 0,
    fiveHourResetsAt: j.five_hour?.resets_at,
    weeklyResetsAt: j.seven_day?.resets_at,
  };
}

export function usageColor(pct: number): string {
  if (pct < 60) return "#34c759"; // green
  if (pct < 85) return "#ff9500"; // amber
  return "#ff3b30"; // red
}
